### PR TITLE
Expand README

### PR DESCRIPTION
**Pull request by seagreen**

### CODE REVIEW

**Comment by seagreen** [comment]
Let me know what you think. I tried to write big, knowing if I overdid it you'd tell me.

File: README.md [hunk 1]
```diff
@@ -2,11 +2,41 @@
...
+
+[Concur](https://github.com/ajnsit/concur): a UI framework that generalizes the Elm architecture. It's based around `Widget`s, sections of the UI that may return a value.
+
+## Upsides and downsides
+
+This strategy is well understood due to its use by [Phoenix LiveView](https://hexdocs.pm/phoenix_live_view) (among other projects).
+
+There are two situations where you should rule `concur-replica` out completely:
+
++ Instant feedback is required
```
Comment (pkamenarsky) [inline README.md]: I'd qualify that a bit - the feedback is more or less dependent on the ping between server and client. Also, in many situations (i.e. clicking on a button), it's pretty hard to tell the difference between 1ms and 50ms. Typing is another matter of course (i.e. client side prediction in replica, etc).

Maybe clarify a bit & move to the "softer costs" section?

File: README.md [hunk 2]
```diff
@@ -2,11 +2,41 @@
...
+
++ It needs to work offline
+
+The strategy also has two softer costs:
+
++ Animations costs bandwidth as well as CPU
```
Comment (pkamenarsky) [inline README.md]: Probably good to qualify as well - css animations are more or less "zero cost" (e.g. update animated css class at start and end of animation). However, continuously updating attributes 20 times a second is not a good idea.

File: README.md [hunk 3]
```diff
@@ -2,11 +2,41 @@
...
+
++ It puts a heavier load on your server than a traditional stateless webapp
+
+Additionally, `concur-replica` is alpha software so it currently has additional deficiencies. Check the issue tracker for these, and they'll be fixed as it matures.
+
+Despite these caveats however, `concur-replica` has *incredible* benefits. No JS! No REST API layer-- you can call whatever Haskell functions you want directly from your UI code. `concur` is an awesome way of making UIs. And you have the Haskell classics: STM is absolutely unfair for building collaborative apps, and the type system shines since UIs are so annoying to test.
+
+We may have a real path to making Haskell a best-in-class UI language.
```
Comment (pkamenarsky) [inline README.md]: Sounds a bit unjustified - i.e. it's still impossible to even build a Google Maps clone efficiently (because we'd need to continuously send `onMouseMove` events to the server and wait on the diff), not to mention that with higher pings something like that would be completely unusable.

Maybe better to remove completely?

File: README.md (hunk 2 above)
Comment (seagreen) [reply README.md]: How would you like to phrase it? "Non-CSS animations" / "animations done in the general purpose language"?

Comment (seagreen) [reply README.md]: "Animations costs bandwidth as well as CPU if not done using CSS" ?

File: README.md (hunk 3 above)
Comment (seagreen) [reply README.md]: This is a great suggestion.

File: README.md (hunk 1 above)
Comment (seagreen) [reply README.md]: > Also, in many situations (i.e. clicking on a button), it's pretty hard to tell the difference between 1ms and 50ms.

Totally! A lot of the time I don't notice the lag with this strategy. However, internet is often choppy and you'll have spikes of more serious lag. I guess this kind of blends into the second point-- a good offline experience can mean "works entirely offline", but can also mean "you don't notice being offline for 1s".

However, I still think this deserves its own bullet point, because developers think about those as two different situations. They may not need Google Maps to work offline, but they still want it to be a good experience with 100ms lag. So whatever bullet point goes here should be the description of why efficient Google Maps is out. Thoughts?

Comment (seagreen) [reply README.md]: Put more clearly: do we have a one sentence way of saying why efficient Google Maps is out?

**Comment by seagreen** [comment]
Haha, I don't think those were nitpicks. Accuracy becomes more important the more hype one adds to a document.

**Comment by seagreen** [comment]
I removed the "best in class" line. Also re-ordered the last clauses of the proceeding paragraph so it ends on a positive note.

File: README.md (hunk 1 above)
Comment (pkamenarsky) [reply README.md]: Yep, agree!

How about "No client side logic is required"? I'd say that disqualifies Google Maps but also *doesn't* more traditional SPAs with lots of buttons and other such comparable elements.

In general I'd say, if Rails (without JS) can do it, Replica can as well. So maybe flip it around and say "Server side logic is sufficient" or something along those lines?

File: README.md (hunk 2 above)
Comment (pkamenarsky) [reply README.md]: I'd say that ties into the previous point - i.e. everything that requires a stream of events upstream (Google Maps) or downstream (Google Maps, custom animations) is out. I think if we go with "No client side logic is required" or something similar that'd cover the animations as a sub-case as well.

File: README.md (hunk 1 above)
Comment (seagreen) [reply README.md]: I think the preceding "How?!" section already describes that concur-replica is a server-side only framework, so saying it again here isn't necessary (if that isn't clear we should expand the "How?!" section).

What I'd like to convey here is the consequences of that.

The streaming examples I actually see as a "soft" cost. The Phoenix LiveView people do use it for animations, eg this https://elixir-doom-fire.herokuapp.com/ (slightly lol example) Whether or not you'd use concur-replica for an app with animations depends-- if it's a personal blog showing how to use a haskell animation library it might be a great choice, if it's meant to have millions of users it's a bad one.

That's why I see the bandwidth issue as a "soft" cost, a situational one.

However, the instant feedback issue is a hard one. No matter how big a server you buy, using concur-replica instead of client side JS will always increase the amount of time until the user gets feedback. Even if you're willing to pay a high server and bandwidth bill, you can't get around this issue. That's why I see it as the "hard" cost.

Additional thoughts:

1. I added another commit mentioning in the section below that both events AND updates cost bandwidth, and moving the mention of "animations" into a parenthetical example.

2. I'm finding this conversation useful beyond just this PR, because I'm learning more about this strategy of doing a UI.

3. I've probably reached the depth of my knowledge on this issue, and won't have too much more to add. Want to take over this PR, reword it to your satisfaction, and then merge it with my blessing? We can always come back to it later once we've had more time to think about it. Also, if you do this I think the line *"if Rails (without JS) can do it, Replica can as well"* might be worth trying to work in, I think that summarizes things well.

Comment (pkamenarsky) [reply README.md]: Good points, and I agree.

I did some minor adjustments, please look them over when you have the time and I'll merge 👍

Comment (seagreen) [reply README.md]: I'm :100: behind those changes, go for it!
